rustls/src/msgs/handshake.rs

```diff
@@ -1190,6 +1192,48 @@ impl HelloRetryRequest {
             _ => None,
         }
     }
+
+    pub(crate) fn ech(&self) -> Option<&Vec<u8>> {
+        let ext = self.find_extension(ExtensionType::EncryptedClientHello)?;
+        match *ext {
+            HelloRetryExtension::EchHelloRetryRequest(ref ech) => Some(ech),
+            _ => None,
+        }
+    }
+
+    fn payload_encode(&self, bytes: &mut Vec<u8>, purpose: Encoding) {
+        self.legacy_version.encode(bytes);
+        HELLO_RETRY_REQUEST_RANDOM.encode(bytes);
+        self.session_id.encode(bytes);
+        self.cipher_suite.encode(bytes);
+        Compression::Null.encode(bytes);
+
+        match purpose {
+            // Standard encoding encodes extensions as they appear.
+            Encoding::Standard => {
+                self.extensions.encode(bytes);
+            }
+            // For the purpose of ECH confirmation, the Encrypted Client Hello extension
+            // must have its payload replaced by 8 zero bytes.
+            //
+            // See draft-ietf-tls-esni-18 7.2.1:
+            // <https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-18#name-sending-helloretryrequest-2>
+            Encoding::EchConfirmation => {
+                let extensions = LengthPrefixedBuffer::new(ListLength::U16, bytes);
```

Comment:
Style nit: maybe take this out of the match arm?

Reply:
I thought that would be workable but it ends up creating two mutable borrows of `bytes` to lift the `LengthPrefixedBuffer` up. The `Standard` arm takes a mut borrow of `bytes` for `self.extensions.encode(bytes)`.